tensorflow_probability/examples/jupyter_notebooks/Cox_Process_TFP.ipynb

```diff
@@ -0,0 +1,1409 @@
+        "        X2: tf.Tensor.\n",
+        "            N x D matrix of N data points each with D features.\n",
+        "        lengthscale: tf.Tensor.\n",
+        "            Lengthscale parameter, a positive scalar or D-dimensional vector.\n",
+        "        variance: tf.Tensor.\n",
+        "            Output variance parameter, a positive scalar.\n",
+        "    \"\"\"\n",
+        "    lengthscale = tf.convert_to_tensor(lengthscale)\n",
+        "    variance = tf.convert_to_tensor(variance)\n",
+        "    dependencies = [tf.assert_positive(lengthscale),\n",
+        "                    tf.assert_positive(variance)]\n",
+        "    lengthscale = control_flow_ops.with_dependencies(dependencies, lengthscale)\n",
+        "    variance = control_flow_ops.with_dependencies(dependencies, variance)\n",
+        "\n",
+        "    X = tf.convert_to_tensor(X)\n",
+        "    X = X / lengthscale\n",
+        "    Xs = tf.reduce_sum(tf.square(X), 1)\n",
+        "    if X2 is None:\n",
+        "        X2 = X\n",
+        "        X2s = Xs\n",
+        "    else:\n",
+        "        X2 = tf.convert_to_tensor(X2)\n",
+        "        X2 = X2 / lengthscale\n",
+        "        X2s = tf.reduce_sum(tf.square(X2), 1)\n",
+        "\n",
+        "    square = tf.reshape(Xs, [-1, 1]) + tf.reshape(X2s, [1, -1]) - \\\n",
+        "        2 * tf.matmul(X, X2, transpose_b=True)\n",
+        "    output = variance * tf.exp(-square / 2)\n",
+        "    return output\n"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "metadata": {
+        "id": "mFOmMDog1_eJ",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "Now we can get to defining our model. We're going to define our Distribution for our point process, $f$, as a multivariate normal distribution. From here, we feed random samples from that into a Poisson Distribution"
+      ]
+    },
+    {
+      "metadata": {
+        "id": "jsgradHqRlWp",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "cell_type": "code",
+      "source": [
+        "# Form (N, V, V) covariance, one matrix per data point.\n",
+        "K = tf.stack([rbf(tf.reshape(xn, [V, 1])) + tf.diag(tf.fill([V], 1e-6))\n",
+        "                for xn in tf.unstack(x_data)])\n",
+        "\n",
+        "# Creating our Covariance Matrix\n",
+        "f = tfd.MultivariateNormalTriL(loc=tf.zeros([N, V]),\n",
+        "                               scale_tril=tf.cholesky(K))\n",
+        "\n",
+        "# Feeding the latent function into a Poisson Distribution\n",
+        "x_init = tfd.Poisson(rate=tf.exp(f.sample()))\n",
+        "x_model = tfd.Poisson(rate=tf.exp(f.sample()))\n",
+        "\n",
+        "# Getting our initial parametrization tensor for setting up the \n",
+        "# trainable_distribution\n",
+        "x_ = evaluate(x_init.sample())"
```

Comment:
For graph mode, it would be preferable to defer evaluation unless you need a fixed sample for each optimization step.
For eager mode, the optimization code in the next cell is not going to work, because it does not run under tf.GradientTape.